lib/modules/manager/gomod/artifacts.ts

```diff
@@ -190,8 +190,20 @@ export async function updateArtifacts({
       );
     }
   }
+
+  const goMod = getGoConfig(newGoModContent);
   const goConstraints =
-    config.constraints?.go ?? getGoConstraints(newGoModContent);
+    config.constraints?.go ?? goMod.toolchain ?? goMod.constraints;
+
+  const flags: string[] = [];
+
+  if (semver.satisfies(goMod.go, '^1.21.0')) {
+    flags.push(`toolchain@${goMod.toolchain ?? 'none'}`);
+  }
+
+  if (!semver.satisfies(goMod.go, '^1.17.0')) {
```

Comment:
```suggestion
  if (!semver.satisfies(goMod.go, '>=1.17.0')) {
```

should also allow go v2